components/vision.py

```diff
@@ -1,36 +1,115 @@
+
+    __slots__ = ("distance", "angle", "timestamp")
 
 
 class Vision:
+    @property
+    def ping_time(self) -> float:
+        return self.ping_time_entry.getDouble(0.0)
+
+    @ping_time.setter
+    def ping_time(self, value: float) -> None:
+        self.ping_time_entry.setDouble(value)
+
+    @property
+    def rio_pong_time(self) -> float:
+        return self.rio_pong_time_entry.getDouble(0.0)
+
+    @property
+    def raspi_pong_time(self) -> float:
+        return self.raspi_pong_time_entry.getDouble(0.0)
+
+    @property
+    def latency(self) -> float:
+        return self.latency_entry.getDouble(0.0)
+
+    @latency.setter
+    def latency(self, value: float) -> None:
+        self.latency_entry.setDouble(value)
+
+    @property
+    def processing_time(self) -> float:
+        return self.processing_time_entry.getDouble(0.0)
+
+    @processing_time.setter
+    def processing_time(self, value: float) -> None:
+        self.processing_time_entry.setDouble(value)
+
     def __init__(self) -> None:
+        self.last_pong = time.monotonic()
+
         self.nt = NetworkTables
         self.table = self.nt.getTable("/vision")
-        self.entry = self.table.getEntry("data")
-        self.lastTime = 0
-        self.repeats = 0
-        self.data = [None, None, None]
+        self.vision_data_entry = self.table.getEntry("data")
+        self.ping_time_entry = self.table.getEntry("ping")
+        self.rio_pong_time_entry = self.table.getEntry("rio_pong")
+        self.raspi_pong_time_entry = self.table.getEntry("raspi_pong")
+        self.latency_entry = self.table.getEntry("clock_offset")
+        self.processing_time_entry = self.table.getEntry("processing_time")
+        self.vision_data = None
 
-    def get_vision_data(self) -> Tuple[float, float, float]:
-        """Returns a tuple containing the distance (metres), angle (radians), and timestamp (time.monotonic)
-        If it can't get the entry, it returns [None, None, None]
+    def get_data(self) -> Optional[VisionData]:
+        """Returns the latest vision data.
 
-        If the pi vision cant see the target it will not send to network tables, meaning this will continue
-        to return the most recently seen target position and the time it was seen
-        """
-        self.data = self.entry.getDoubleArray([None, None, None])
-        return self.data
+        Returns None if there is no vision data.
+         """
+        return self.vision_data
 
-    def is_ready(self) -> bool:
-        if self.entry.exists() and self.data[2] != None:
-            if self.data[2] == self.lastTime:  # if it gets the same time twice
-                self.repeats += 1
-            else:
-                self.repeats = 0
-            self.lastTime = self.data[2]
+    def execute(self) -> None:
+        data = None
+        data = self.vision_data_entry.getDoubleArray(None)
+        if data is not None:
+            self.vision_data = VisionData(*data)
+            self.ping()
```

Comment:
The rest of this method would be delegated to the comms class method. No knowledge of this stuff need be in the Vision class. Vision should not know about networking or ping-pong protocols, etc.